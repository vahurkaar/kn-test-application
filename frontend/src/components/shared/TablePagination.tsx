import { Pagination } from 'react-bootstrap';
import React from 'react';
import Page from '../../models/Page';
import { observer } from 'mobx-react';

export interface TablePaginationProps {
  page: Page,
  onSelect: (page: Page) => void;
}

const TablePagination = observer(({ page, onSelect }: TablePaginationProps) => {
  const currentPage = page.number + 1;
  const goToPage = (number: number) => onSelect({ ...page, number });
  const goToFirstPage = () => goToPage(0);
  const goToLastPage = () => goToPage(page.totalPages - 1);
  const goToNext = () => goToPage(page.number + 1);
  const goToPrevious = () => goToPage(page.number - 1);

  const isNextEnabled = page.number < page.totalPages - 1;
  const isPrevEnabled = page.number > 0;

  const renderPreviousItems = (currentPage: number, maxItems: number) => {
    const items = [];
    for (let pageNumber = Math.max(1, currentPage - maxItems); pageNumber < currentPage; pageNumber++) {
      items.push(<Pagination.Item data-testid={`item-${pageNumber}`} key={`prev-${pageNumber}`} onClick={() => goToPage(pageNumber - 1)}>{pageNumber}</Pagination.Item>);
    }

    return items;
  }

  const renderNextItems = (currentPage: number, maxItems: number) => {
    const items = [];
    for (let pageNumber = Math.min(page.totalPages, currentPage + maxItems); pageNumber > currentPage; pageNumber--) {
      items.push(<Pagination.Item data-testid={`item-${pageNumber}`} key={`next-${pageNumber}`} onClick={() => goToPage(pageNumber - 1)}>{pageNumber}</Pagination.Item>);
    }

    items.reverse();
    return items;
  }

  return (
    <Pagination size="sm" className="justify-content-center">
      <Pagination.First data-testid="first" disabled={!isPrevEnabled} onClick={goToFirstPage} />
      <Pagination.Prev data-testid="prev" disabled={!isPrevEnabled} onClick={goToPrevious} />

      {renderPreviousItems(currentPage, 3)}
      <Pagination.Item data-testid={`item-${currentPage}`} active>{currentPage}</Pagination.Item>
      {renderNextItems(currentPage, 3)}

      <Pagination.Next data-testid="next" disabled={!isNextEnabled} onClick={goToNext} />
      <Pagination.Last data-testid="last" disabled={!isNextEnabled} onClick={goToLastPage} />
    </Pagination>
  );
});

export default TablePagination;