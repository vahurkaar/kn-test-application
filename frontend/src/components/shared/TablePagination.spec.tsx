import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import TablePagination, { TablePaginationProps } from './TablePagination';

describe('TablePagination', () => {
  const create = (props?: Partial<TablePaginationProps>) => {
    const paginationProps = {
      page: { number: 0, totalPages: 1, totalElements: 10, size: 10 },
      onSelect: jest.fn(),
      ...props
    }
    return render(<TablePagination {...paginationProps} />);
  };

  describe('render', () => {
    it('should render with 1 page', () => {
      const { container } = create();
      expect(screen.getAllByRole('listitem')).toHaveLength(5);
      expect(container).toMatchSnapshot();
    });

    it('should render first page and 3 next pages when result has 5 pages', () => {
      const { container } = create({ page: { number: 0, totalPages: 5, totalElements: 50, size: 10 } });
      expect(screen.getAllByRole('listitem')).toHaveLength(8);
      expect(container).toMatchSnapshot();
    });

    it('should render second page with 1 previous and 3 next pages when result has 5 pages', () => {
      const { container } = create({ page: { number: 1, totalPages: 5, totalElements: 50, size: 10 } });
      expect(screen.getAllByRole('listitem')).toHaveLength(9);
      expect(container).toMatchSnapshot();
    });

    it('should render second page with 1 previous and 2 next pages when result has 4 pages', () => {
      const { container } = create({ page: { number: 1, totalPages: 4, totalElements: 40, size: 10 } });
      expect(screen.getAllByRole('listitem')).toHaveLength(8);
      expect(container).toMatchSnapshot();
    });

    it('should render fifth page with 3 previous and 3 next pages when result has 10 pages', () => {
      const { container } = create({ page: { number: 4, totalPages: 10, totalElements: 100, size: 10 } });
      expect(screen.getAllByRole('listitem')).toHaveLength(11);
      expect(container).toMatchSnapshot();
    });

    it('should render last page with 3 previous when result has 10 pages', () => {
      const { container } = create({ page: { number: 9, totalPages: 10, totalElements: 100, size: 10 } });
      expect(screen.getAllByRole('listitem')).toHaveLength(8);
      expect(container).toMatchSnapshot();
    });

    it('should render second to last page with 3 previous and 1 next when result has 10 pages', () => {
      const { container } = create({ page: { number: 8, totalPages: 10, totalElements: 100, size: 10 } });
      expect(screen.getAllByRole('listitem')).toHaveLength(9);
      expect(container).toMatchSnapshot();
    });
  });

  describe('onSelect', () => {
    it('should open last page', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 0, totalPages: 10, totalElements: 100, size: 10 }, });

      const lastPageItem = screen.getByTestId("last");
      fireEvent.click(lastPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 9, totalPages: 10, totalElements: 100, size: 10 });
    });

    it('should open next page', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 0, totalPages: 10, totalElements: 100, size: 10 }, });

      const nextPageItem = screen.getByTestId("next");
      fireEvent.click(nextPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 1, totalPages: 10, totalElements: 100, size: 10 });
    });

    it('should open next page directly', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 0, totalPages: 10, totalElements: 100, size: 10 }, });

      const nextPageItem = screen.getByTestId("item-2");
      fireEvent.click(nextPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 1, totalPages: 10, totalElements: 100, size: 10 });
    });

    it('should open fourth page', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 0, totalPages: 10, totalElements: 100, size: 10 }, });

      const fourthPageItem = screen.getByTestId("item-4");
      fireEvent.click(fourthPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 3, totalPages: 10, totalElements: 100, size: 10 });
    });

    it('should open first page', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 9, totalPages: 10, totalElements: 100, size: 10 }, });

      const firstPageItem = screen.getByTestId("first");
      fireEvent.click(firstPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 0, totalPages: 10, totalElements: 100, size: 10 });
    });

    it('should open previous page', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 9, totalPages: 10, totalElements: 100, size: 10 }, });

      const previousPageItem = screen.getByTestId("prev");
      fireEvent.click(previousPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 8, totalPages: 10, totalElements: 100, size: 10 });
    });

    it('should open previous page directly', () => {
      const onSelect = jest.fn();

      create({ onSelect, page: { number: 9, totalPages: 10, totalElements: 100, size: 10 }, });

      const previousPageItem = screen.getByTestId("item-8");
      fireEvent.click(previousPageItem);
      expect(onSelect).toHaveBeenCalledWith({ number: 7, totalPages: 10, totalElements: 100, size: 10 });
    });
  });
});