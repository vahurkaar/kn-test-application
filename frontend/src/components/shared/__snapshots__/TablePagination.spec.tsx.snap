// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePagination render should render fifth page with 3 previous and 3 next pages when result has 10 pages 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="first"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="prev"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-2"
        href="#"
        role="button"
        tabindex="0"
      >
        2
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-3"
        href="#"
        role="button"
        tabindex="0"
      >
        3
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-4"
        href="#"
        role="button"
        tabindex="0"
      >
        4
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-5"
      >
        5
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-6"
        href="#"
        role="button"
        tabindex="0"
      >
        6
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-7"
        href="#"
        role="button"
        tabindex="0"
      >
        7
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-8"
        href="#"
        role="button"
        tabindex="0"
      >
        8
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="next"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="last"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`TablePagination render should render first page and 3 next pages when result has 5 pages 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="first"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="prev"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </span>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-1"
      >
        1
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-2"
        href="#"
        role="button"
        tabindex="0"
      >
        2
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-3"
        href="#"
        role="button"
        tabindex="0"
      >
        3
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-4"
        href="#"
        role="button"
        tabindex="0"
      >
        4
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="next"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="last"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`TablePagination render should render last page with 3 previous when result has 10 pages 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="first"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="prev"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-7"
        href="#"
        role="button"
        tabindex="0"
      >
        7
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-8"
        href="#"
        role="button"
        tabindex="0"
      >
        8
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-9"
        href="#"
        role="button"
        tabindex="0"
      >
        9
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-10"
      >
        10
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="next"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="last"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </span>
    </li>
  </ul>
</div>
`;

exports[`TablePagination render should render second page with 1 previous and 2 next pages when result has 4 pages 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="first"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="prev"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-1"
        href="#"
        role="button"
        tabindex="0"
      >
        1
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-2"
      >
        2
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-3"
        href="#"
        role="button"
        tabindex="0"
      >
        3
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-4"
        href="#"
        role="button"
        tabindex="0"
      >
        4
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="next"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="last"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`TablePagination render should render second page with 1 previous and 3 next pages when result has 5 pages 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="first"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="prev"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-1"
        href="#"
        role="button"
        tabindex="0"
      >
        1
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-2"
      >
        2
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-3"
        href="#"
        role="button"
        tabindex="0"
      >
        3
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-4"
        href="#"
        role="button"
        tabindex="0"
      >
        4
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-5"
        href="#"
        role="button"
        tabindex="0"
      >
        5
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="next"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="last"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`TablePagination render should render second to last page with 3 previous and 1 next when result has 10 pages 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="first"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="prev"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-6"
        href="#"
        role="button"
        tabindex="0"
      >
        6
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-7"
        href="#"
        role="button"
        tabindex="0"
      >
        7
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-8"
        href="#"
        role="button"
        tabindex="0"
      >
        8
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-9"
      >
        9
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="item-10"
        href="#"
        role="button"
        tabindex="0"
      >
        10
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="next"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        data-testid="last"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`TablePagination render should render with 1 page 1`] = `
<div>
  <ul
    class="justify-content-center pagination pagination-sm"
  >
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="first"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          class="visually-hidden"
        >
          First
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="prev"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          class="visually-hidden"
        >
          Previous
        </span>
      </span>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        data-testid="item-1"
      >
        1
        <span
          class="visually-hidden"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="next"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        data-testid="last"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          class="visually-hidden"
        >
          Last
        </span>
      </span>
    </li>
  </ul>
</div>
`;
